### PR TITLE
Made `sample` type agnostic and GPU compatible

### LINKED ISSUE #92: [BUG] `sample` in `mps.jl` broken when using CUDA
The following minimal working example crashes in a `MethodError: no method matching DenseVector{Float64}(::UndefInitializer, ::Tuple{Int64})` in ITensorMPS version 0.3.1:

```julia
using CUDA
using ITensors, ITensorMPS, NDTensors

N = 10
sites = siteinds("Qubit", N)
ψ = NDTensors.cu(MPS(sites, repeat(["+"], N)))

sample!(ψ)
```

The error can be mitigated by copying the MPS to the cpu first, or by adopting the `sample` function in the following way:
```julia
function sample(rng::AbstractRNG, m::MPS)
  N = length(m)

  if orthocenter(m) != 1
    error("sample: MPS m must have orthocenter(m)==1")
  end
  if abs(1.0 - norm(m[1])) > 1E-8
    error("sample: MPS is not normalized, norm=$(norm(m[1]))")
  end

  result = zeros(Int, N)
  A = m[1]

  for j in 1:N
    s = siteind(m, j)
    d = dim(s)
    # Compute the probability of each state
    # one-by-one and stop when the random
    # number r is below the total prob so far
    pdisc = 0.0
    r = rand(rng)
    # Will need n,An, and pn below
    n = 1
    An = NDTensors.cu(ITensor()) # this line is changed
    pn = 0.0
    while n <= d
      projn = ITensor(s)
      projn[s => n] = 1.0
      An = A * dag(NDTensors.cu(projn)) # this line is changed
      pn = real(scalar(dag(An) * An))
      pdisc += pn
      (r < pdisc) && break
      n += 1
    end
    result[j] = n
    if j < N
      A = m[j + 1] * An
      A *= (1.0 / sqrt(pn))
    end
  end
  return result
end
```

I suppose that there is a more performant fix, since the above implementation is rather slow.

### PR DESCRIPTION
**Pull request by tipfom**
Fixed the issue of `sample` mentioned in #92.

Closes #92.

### CODE REVIEW

File: src/mps.jl [hunk 1]
```diff
@@ -655,6 +655,8 @@ function sample(rng::AbstractRNG, m::MPS)
     error("sample: MPS is not normalized, norm=$(norm(m[1]))")
   end
 
+  ElT = promote_itensor_eltype(m)
```
Comment (mtfishman) [inline src/mps.jl:658]: I've been preferring `scalartype(m)` for this these days, probably we will remove `promote_itensor_eltype`.

**Comment by kmp5VT** [comment]
@mtfishman I can look around other code in `ITensorMPS` to see if there is more implicit `Float64` typing and open another PR if there is. Thank you @tipfom for your PR!

**Comment by mtfishman** [comment]
Thanks @kmp5VT. Also it would be great to start switching some tests over to being able to test on GPU, looping over device backends the same way we do in the NDTensors/ITensors tests. For now it could just test using JLArrays, but ultimately we should also set things up with Jenkins to run the tests on GPUs.

**Comment by tipfom** [comment]
@mtfishman I've replaced all occurances of `promote_itensor_eltype` by `scalartype`. Hope that's okay; sorry for the delay.

**Comment by mtfishman** [comment]
> @mtfishman I've replaced all occurances of `promote_itensor_eltype` by `scalartype`. Hope that's okay; sorry for the delay.

Thanks!

**Pull request merged**